### ISSUE TITLE
Testing FastDifferentiation.jl for TORA.jl

### ISSUE THREAD

**Issue by ferrolho** [brianguenter/FastDifferentiation.jl#33]
Hi, Brian! 👋 I am currently using [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl) within [TORA.jl](https://github.com/JuliaRobotics/TORA.jl) for computing Jacobians of user-defined vector-valued functions. However, I noticed that [FastDifferentiation.jl](https://github.com/brianguenter/FastDifferentiation.jl) (this package) may provide quicker Jacobian calculations compared to [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl), and so I would like to give it a go.

I've created a Jupyter notebook, [TestFastDifferentiation.ipynb](https://github.com/JuliaRobotics/TORA.jl/blob/hf/test-FastDifferentiation.jl/notebooks/TestFastDifferentiation.ipynb), on a new branch (`hf/test-FastDifferentiation.jl`), where I want to benchmark [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl) and [FastDifferentiation.jl](https://github.com/brianguenter/FastDifferentiation.jl) side-by-side. I've managed to create the variables with
```julia
x = make_variables(:x, robot.n_q + robot.n_v + robot.n_τ + robot.n_q + robot.n_v)
```
but I can't figure out the correct syntax for the actual evaluation of the Jacobian.

P.s. the function I am trying to differentiate is [`forward_dynamics_defects!`](https://github.com/JuliaRobotics/TORA.jl/blob/42a3aef6ae6fb6ef9d1ff3bf14ba15cec7fc22f2/src/constraints/dynamics.jl#L11-L47).

**Comment by brianguenter** [comment]
I'm not entirely sure what you intend to take the derivative of. I don't see a call to `FastDifferentiation.jacobian()` anywhere in the code you linked. 

The standard way to use FastDifferentiation is:

* define variables using `make_variables`
* pass these variables to the function you want to differentiate, in place of the floating point arguments you'd normally pass to that function. The element type returned by calling your function this way will be `FastDifferentiation` symbolic expression graphs, which will display as type `Node`.
* call `jacobian` on the result of this function call. This returns a matrix of `FastDifferentiation` symbolic expression graphs.
* pass the result from `jacobian` as the first argument to `make_function.`
* call the function created with `make_function` with floating point arguments.

Example:
```julia
julia> x = make_variables(:x,2)
2-element Vector{FastDifferentiation.Node}:
 x1
 x2

julia> f(z) = z .* z
f (generic function with 1 method)

julia> symb_f = f(x)
2-element Vector{FastDifferentiation.Node{typeof(*), 2}}:
 (x1 * x1)
 (x2 * x2)

julia> jac_f = jacobian(symb_f,x)
2×2 Matrix{FastDifferentiation.Node}:
 (x1 + x1)        0.0
       0.0  (x2 + x2)

julia> jac_func = make_function(jac_f,x)
RuntimeGeneratedFunction(#=in FastDifferentiation=#, #=using FastDifferentiation=#, :((input_variables,)->begin
          #= c:\Users\seatt\Source\FastDifferentiation.jl\src\CodeGeneration.jl:93 =#
          #= c:\Users\seatt\Source\FastDifferentiation.jl\src\CodeGeneration.jl:93 =# @inbounds begin
                  #= c:\Users\seatt\Source\FastDifferentiation.jl\src\CodeGeneration.jl:94 =#
                  begin
                      result = zeros(promote_type(Float64, eltype(input_variables)), 2, 2)
                      begin
                          var"##293" = input_variables[1] + input_variables[1]
                          result[CartesianIndex(1, 1)] = var"##293"
                      end
                      begin
                          var"##294" = input_variables[2] + input_variables[2]
                          result[CartesianIndex(2, 2)] = var"##294"
                      end
                      return result
                  end
              end
      end))

julia> jac_func([2.0,3.0])
2×2 Matrix{Float64}:
 4.0  0.0
 0.0  6.0
```

If you explain what function you are trying to differentiate I can help you figure out how.

**Comment by brianguenter** [comment]
Is this what you intended:
```
function test_robot()
    robot = TORA.create_robot_franka("panda_arm", vis)
    x = make_variables(:x, robot.n_q + robot.n_v + robot.n_τ + robot.n_q + robot.n_v)
    defects = zeros(robot.n_q + robot.n_v)
    dyn = TORA.forward_dynamics_defects!(defects, robot, x, dt)
    jac_defects = jacobian(dyn, x)
    dyn_func = make_function(jac_defects, x)
    dyn_func(rand(robot.n_q + robot.n_v + robot.n_τ + robot.n_q + robot.n_v))
end
```
If this looks correct I'll get this working. There is currently a type promotion error in the `muladd` function that might take a little while to fix.

**Comment by ferrolho** [comment]
Yes, this looks correct. Thanks!

I am not sure if the `muladd` issue you are referring to is the same as this one, but I get
```julia
MethodError: no method matching AbstractFloat(::FastDifferentiation.Node{Symbol, 0})
```
at the line
```julia
dyn = TORA.forward_dynamics_defects!(defects, robot, x, dt)
```

**Comment by brianguenter** [comment]
Yeah, I think that is a separate error related to sincos. This function wasn't defined for Node types because SymbolicUtils.number_methods doesn't define it. I've fixed it in my local code, not sure if that has been pushed to the repo yet. But the promotion error is still there. Have to attend to other things for a few hours but should have this fixed by tonight.

**Comment by ferrolho** [comment]
The issue has been closed, but I just tested this on the latest main and there is this error:
```
promotion of types FastDifferentiation.Node{Int64, 0}, FastDifferentiation.Node{Float64, 0} and FastDifferentiation.Node{Int64, 0} failed to change any arguments
```

**Comment by brianguenter** [comment]
Fixed the type promotion error, which exposed problems in the LinearAlgebra code. 

Your code calls dynamics_solve! which calls LU which has conditionals which FastDifferentiation.jl doesn't yet support. 

Symbolics.jl has an LU factorization which uses a heuristic to do partial pivoting since the value of the matrix elements is not known at the time the symbolic expression is created. Instead of comparing numeric values they compare the number of terms in the symbolic expressions. No idea how well this heuristic works or whether this would result in good pivot choices.

I can probably implement this symbolic LU in FastDifferentiation. Would that be useful?

**Comment by brianguenter** [comment]
To make this work the signature of the symbolic LU would have to exactly match the signature of the LU you are calling, except for the element type of the arrays.

**Comment by brianguenter** [comment]
I think it should be possible to overload the \ operator specialized to matrices of Node type. Then it would be completely transparent. 

The partial pivoting would be done symbolically in the LU factorization rather than based on actual numeric values of matrix elements ,which might in some cases lead to slightly worse condition numbers than you would expect. Not sure if this will be a big problem.

Other people have needed LU factorization as well so it's worth doing.

**Comment by baggepinnen** [comment]
The `\` matrix operator appears in MTK systems compiled with JuliaSimCompiler as well, differentiating through them efficiently has proven rather difficult (especially if they represent multibody systems)

**Comment by brianguenter** [comment]
I think I have figured out how to make FastDifferentiation work with ChainRules. I'm currently working on conditionals, so adding ChainRules support will come once I complete conditionals. ChainRules is tricky because FastDifferentiation is inherently a scalar system now and ChainRules is all about Jacobian vector products, which is not really a thing in FastDifferentiation.
